packages/monitor-v2/src/llm-bot/common.ts

```diff
@@ -63,27 +92,94 @@ export class OptimisticOracleRequest {
     this.resolvedValue = data.resolvedValue;
     this.resolveTx = data.resolveTx;
     this.disputeTx = data.disputeTx;
+    this.blockNumber = data.blockNumber;
+    this.transactionIndex = data.transactionIndex;
+  }
+
+  getId(): string {
+    return calculateRequestId(this.body, this.identifier, this.timestamp, this.requester);
+  }
+
+  setProposer(proposer: string): void {
+    this.proposer = proposer;
+  }
+
+  setProposedValue(proposedValue: number | boolean): void {
+    this.proposedValue = proposedValue;
+  }
+
+  setProposeTx(proposeTx: string): void {
+    this.proposeTx = proposeTx;
+  }
+
+  setDisputableUntil(disputableUntil: number): void {
+    this.disputableUntil = disputableUntil;
+  }
+
+  setResolvedValue(resolvedValue: number | boolean): void {
+    this.resolvedValue = resolvedValue;
+  }
+
+  setResolveTx(resolveTx: string): void {
+    this.resolveTx = resolveTx;
+  }
+
+  setDisputeTx(disputeTx: string): void {
+    this.disputeTx = disputeTx;
+  }
+
+  setBlockNumber(blockNumber: number): void {
+    this.blockNumber = blockNumber;
+  }
+
+  setTransactionIndex(transactionIndex: number): void {
+    this.transactionIndex = transactionIndex;
   }
 }
 
+const EMPTY_BLOCK_RANGE: BlockRange = [0, 0];
+
 /**
  * Abstract class representing a client to interact with an Optimistic Oracle and store the requests.
  */
 export abstract class OptimisticOracleClient<R extends OptimisticOracleRequest> {
   protected provider: Provider;
-  protected requests: R[];
-  protected fetchedBlockRanges: [number, number][];
+  readonly requests: Map<string, R>;
+  protected fetchedBlockRanges: BlockRange[];
 
   /**
    * Constructs a new instance of OptimisticOracleClient.
    * @param _provider The provider used for interacting with the blockchain.
-   * @param _requests (Optional) The list of Optimistic Oracle requests.
+   * @param _requests (Optional) The map of Optimistic Oracle requests.
    * @param _fetchedBlockRanges (Optional) The block ranges of the fetched requests.
+   * @dev requests are stored in a map for faster access and to avoid duplicates.
    */
-  protected constructor(_provider: Provider, _requests?: R[], _fetchedBlockRanges?: [number, number][]) {
+  protected constructor(
+    _provider: Provider,
+    _requests: Map<string, R> = new Map(),
+    _fetchedBlockRanges: BlockRange[] = [EMPTY_BLOCK_RANGE]
+  ) {
     this.provider = _provider;
-    this.requests = _requests || [];
-    this.fetchedBlockRanges = _fetchedBlockRanges ? _fetchedBlockRanges : [];
+    this.requests = _requests;
+    this.fetchedBlockRanges = _fetchedBlockRanges;
+  }
+
+  /**
+   * Returns a copy of the requests.
+   * @returns A copy of the requests.
+   * @dev This is a deep copy.
+   */
+  getRequestsCopy(): Map<string, R> {
```

Comment:
Curious about this design decision: if the request, itself, is immutable, then I would expect that a deep copy shouldn't ever be necessary, since the same request objects could be used repeatedly and only replaced individually when modified. Thoughts?

Reply:
What made me make this decision was enforcing the immutability of a OOv2 client in a more intuitive way for the developer. Request objects should be updated during their life cycle rather than duplicated. Some properties are now settable for this purpose (see your comment above). Handling the copy in the the [base class](https://github.com/UMAprotocol/protocol/blob/b04e25776ab034129b389317e9e005aaeeb3cb24/packages/monitor-v2/src/llm-bot/common.ts#L237) seems better than in the implementing class so the later only needs to update it's requests and don't care about overriding anything.

You are suggesting to reuse the Requests and replace them instead. If we do this we still need to do a copy of the previous Request at some point so we don't override them right? The problem with this is that it still requires to add logic to handle this Requests updates without overriding - copy `Request` instead of copy `requests`. Adding this selective copy logic in the base class seems more cumbersome than a deepcopy of the requests. Alternatively it could be the implementation class the one responsible for not overriding props of a previous Request and instead doing a copy but this would be more error prone.

Did I understood your concerns?

Reply:
Here's my rough thought on how this can be done: https://github.com/UMAprotocol/protocol/commit/8b2e5460202b7f40c3064a42c69d555d8113606c. Generally, I think the request object being immutable by definition makes them super easy to pass around and work with, especially concurrently. Preemptive deep copying does mean that on the creation of a new client, the requests won't be reused, but it doesn't really guarantee much about how the request will act out in the wild, once it's passed to other elements of the system. In this approach, I tried to get rid of any notion of modifying or setting fields, while keeping it fairly easy to do copy-modify operations without much syntactical overhead.

A few things:
1. The request is immutable, but the fields are grouped in the data based on being set _together_. When we have the proposer, we should also have the proposal txn hash, for instance. This means if the proposal data changes, but the original request data doesn't, then the reference will point to the same request data with no copies. Only the proposal data reference will change to the updated one.
2. We still copy the map, but in a shallow way, which makes it less intensive. Similar to the above, new references but no copied data.
3. When updating a request, we just add in the "group" of data that we want to amend and a new request pops out. So you might call `const newRequest = request.update( { proposalData: { proposer, ... } } );` I think this is syntactically pretty simple to do in an update function, similar to just setting fields in an existing object.

Thoughts on this approach? Note: it's just a rough idea, not saying it must be done this way. It seems intuitive to me, but let me know if you think I'm wrong!!

Reply:
Thanks @mrice32,
I like your proposal, it makes sense and it's a good alternative to deepcopy! Thanks for your suggestion. I've added your commit and fixed some small issues.